### PR TITLE
Move python-independent c10d implementations to torch/lib

### PR DESCRIPTION
**Pull request by gmagogsfm**
* This is a pre-step to build c10d into libtorch
* Includes a minor cleanup in c10d/CMakeLists.txt

### CODE REVIEW

File: torch/lib/c10d/frontend.cpp [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#include <torch/csrc/distributed/c10d/frontend.h>
+#include <c10d/frontend.h>
```
Comment (wanchaol) [inline torch/lib/c10d/frontend.cpp:1]: nit: change to frontend.hpp

**Review by wanchaol — APPROVED** [review]
code move looks good to me, cc @mrshenli

**Review by mrshenli — APPROVED** [review]
file reorgs LGTM! Stamp to unblock, but please wait for all OSS tests. 

qq: I might lost some context from our last discussion. any reason for moving `csrc/distributed/c10d` contents into `lib/c10d`, not the other way around?

**Comment by gmagogsfm** [comment]
> file reorgs LGTM! Stamp to unblock, but please wait for all OSS tests.
> 
> qq: I might lost some context from our last discussion. any reason for moving `csrc/distributed/c10d` contents into `lib/c10d`, not the other way around?

lib/c10d files are used for libtorch_cpu/cuda build while csrc/distributed/c10d are used for libtorch_python build. In our last discussion, we talked about moving most of c10d into libtorch_cpu/cuda so that TorchScript can call into them without Python. Let me know if this doesn't align with your impression, happy to chat more.

**Comment by mrshenli** [comment]
Thanks for the explanation. 

Just a heads-up, there might be conflict with #47270. If so, please coordinate landing with @SciPioneer

**Pull request closed without merging**